src/System.Console/src/System/IO/StdInStreamReader.cs

```diff
@@ -0,0 +1,350 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using Microsoft.Win32.SafeHandles;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Text;
+using System.Runtime.InteropServices;
+
+namespace System.IO
+{
+    /* This class is used by for reading from the stdin.
+     * It is designed to read stdin in raw mode for interpreting
+     * key press events and maintain its own buffer for the same.
+     * which is then used for all the Read operations
+     */
+    internal class StdInStreamReader : StreamReader
+    {
+        char[] unprocessedBufferToBeRead; // Buffer that might have already been read from stdin but not yet processed.
+        const int bytesToBeRead = 255; // No. of bytes to be read from the stream at a time.
+        int startIndex; // First unprocessed index in the buffer;
+        int endIndex; // Last unprocessed index in the buffer;
+
+        internal StdInStreamReader(Stream stream, Encoding encoding, bool detectEncodingFromByteOrderMarks, int bufferSize, bool leaveOpen) : base(stream, encoding, detectEncodingFromByteOrderMarks, bufferSize, leaveOpen)
+        {
+            unprocessedBufferToBeRead = new char[encoding.GetMaxCharCount(bufferSize)];
+            startIndex = endIndex = -1; // There is no unprocessed index yet.
+        }
+
+        /// <summary> Checks whether the unprocessed buffer is empty. </summary>
+        bool IsExtraBufferEmpty()
+        {
+            if (startIndex == -1 && endIndex == -1) return true;
+
+            if (startIndex > endIndex) // Everything has been processed;
+            {
+                startIndex = -1;
+                endIndex = -1;
+                return true;
+            }
+
+            return false;
+        }
+
+        /// <summary> Reads the buffer in the raw mode. </summary>
+        unsafe void ReadExtraBuffer()
+        {
+            Debug.Assert(IsExtraBufferEmpty());
+
+            byte[] buffer = new byte[bytesToBeRead];
+            fixed (byte* bufPtr = buffer)
```

Comment:
If you declare ReadStdinUnbuffered to take a ```byte[]``` instead of a ```byte*```, you won't need to do this manually.  However (contrary to my previous question/comment), since it's a relatively small buffer and since we're not returning it or passing it to anyone else, you could avoid the heap allocation entirely by using a stackalloc here instead.  We should probably do one or the other, either reusing the buffer across calls or using stackalloc.